Segmentation fault when loading coordinators for empty-string consumer group name

While testing the latest version of [KMinion](https://github.com/redpanda-data/kminion) (a Kafka monitoring tool that uses `franz-go` under the hood) which recently upgraded `franz-go` from version `v1.4.1` to version `v1.10.0`, the application is now crashing with the following panic:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x7de84d]

goroutine 73 [running]:
github.com/twmb/franz-go/pkg/kgo.(*describeGroupsSharder).shard(0x800?, {0xcf0750?, 0xc000460100?}, {0xcf4828?, 0xc00001a9c0}, {0x7f2688649fff?, 0x429ea5?})
        /go/pkg/mod/github.com/twmb/franz-go@v1.10.0/pkg/kgo/client.go:2544 +0xa2d
github.com/twmb/franz-go/pkg/kgo.(*Client).handleShardedReq.func2({0x0, {0xcf4828, 0xc00001a9c0}, {0x0, 0x0}})
        /go/pkg/mod/github.com/twmb/franz-go@v1.10.0/pkg/kgo/client.go:1730 +0x151
github.com/twmb/franz-go/pkg/kgo.(*Client).handleShardedReq(0xc000278000, {0xcf0750?, 0xc000460100}, {0xcf4828?, 0xc00001a9c0})
        /go/pkg/mod/github.com/twmb/franz-go@v1.10.0/pkg/kgo/client.go:1821 +0x9f9
github.com/twmb/franz-go/pkg/kgo.(*Client).shardedRequest(0xc000278000, {0xcf07f8?, 0xc00011ac60?}, {0xcf4828?, 0xc00001a9c0})
        /go/pkg/mod/github.com/twmb/franz-go@v1.10.0/pkg/kgo/client.go:978 +0x691
github.com/twmb/franz-go/pkg/kgo.(*Client).RequestSharded(0xc000494000?, {0xcf07f8?, 0xc00011ac60?}, {0xcf4828?, 0xc00001a9c0?})
        /go/pkg/mod/github.com/twmb/franz-go@v1.10.0/pkg/kgo/client.go:909 +0x3a
github.com/cloudhut/kminion/v2/minion.(*Service).DescribeConsumerGroups(0xc000494000, {0xcf07f8, 0xc00011ac60})
        /app/minion/describe_consumer_groups.go:70 +0x15b
github.com/cloudhut/kminion/v2/prometheus.(*Exporter).collectConsumerGroups(0xc000360000, {0xcf07f8?, 0xc00011ac60?}, 0xcea010?)
        /app/prometheus/collect_consumer_groups.go:18 +0x5e
github.com/cloudhut/kminion/v2/prometheus.(*Exporter).Collect(0xc000360000, 0xc000392f60?)
        /app/prometheus/exporter.go:235 +0x205
github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1()
        /go/pkg/mod/github.com/prometheus/client_golang@v1.14.0/prometheus/registry.go:456 +0x10d
created by github.com/prometheus/client_golang/prometheus.(*Registry).Gather
        /go/pkg/mod/github.com/prometheus/client_golang@v1.14.0/prometheus/registry.go:467 +0x55d
```

I traced back the problem to my Kafka cluster having an empty-string consumer group name. We use Confluent's Kafka Enterprise version and the Control Center component has the (bad) habit of creating these empty-string named consumer groups for itself.

When KMinion attempts to gather information about that group using `franz-go`, then it crashes as shown above. Important is to note that this was not crashing in a previous version of KMinion that used `franz-go` version `v1.4.1`, even with these empty-string name consumer groups in the cluster. `franz-go` was able to fetch metadata correctly for these groups.

I manually upgraded `franz-go` in KMinion to latest `v1.10.3` and the problem is still there.

It seems that the [`loadCoordinators()` function](https://github.com/twmb/franz-go/blob/2e4f905202516652ee08b4d8df0c499afa4a89f2/pkg/kgo/client.go#L1149) in `franz-go` is not playing well when the requested consumer group is an empty name, i.e. the string `""`, and the returned `berr.b` structure becomes `nil` for it.

Original issue opened in KMinion's repository: https://github.com/redpanda-data/kminion/issues/177

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,20 @@
+v1.10.4
+===
+
+This patch release fixes two bugs introduced with v1.10.0. These bugs are not
+encountered in when using the client to simply consume or produce. Only admin
+usages of the client may encounter the bugs this patch is fixing.
+
+v1.10.0 introduced support for batch offset fetching or coordinator finding.
+These changes introduced a bug where empty coordinator keys (i.e., group names
+or transactional IDs) would be stripped from requests, and then a field in a
+nil pointer could be accessed and panic the program. These changes also
+introduced a bug that did not properly mirror one field for batched
+`FindCoordinator` requests.
+
+- [`ca67da4`](https://github.com/twmb/franz-go/commit/ca67da4) **bugfix** kgo: fix batch coordinator fetching
+- [`c6f7f9a`](https://github.com/twmb/franz-go/commit/c6f7f9a) **bugfix** kgo: allow empty groups when finding coordinator / fetching offsets
+
 v1.10.3
 ===
 

--- a/pkg/kgo/client.go
+++ b/pkg/kgo/client.go
@@ -1154,9 +1154,7 @@ func (cl *Client) loadCoordinators(ctx context.Context, typ int8, keys ...string
 
 	toRequest := make(map[string]bool, len(keys)) // true == bypass the cache
 	for _, key := range keys {
-		if len(key) > 0 {
-			toRequest[key] = false
-		}
+		toRequest[key] = false
 	}
 
 	// For each of these keys, we have two cases:
@@ -2324,9 +2322,7 @@ func (cl *offsetFetchSharder) shard(ctx context.Context, kreq kmsg.Request, last
 	}
 	groups := make([]string, 0, len(req.Groups))
 	for i := range req.Groups {
-		if g := req.Groups[i].Group; len(g) > 0 {
-			groups = append(groups, req.Groups[i].Group)
-		}
+		groups = append(groups, req.Groups[i].Group)
 	}
 
 	coordinators := cl.loadCoordinators(ctx, coordinatorTypeGroup, groups...)
@@ -2487,9 +2483,7 @@ func (*findCoordinatorSharder) shard(_ context.Context, kreq kmsg.Request, lastE
 	}
 	req.CoordinatorKeys = req.CoordinatorKeys[:0]
 	for key := range uniq {
-		if len(key) > 0 {
-			req.CoordinatorKeys = append(req.CoordinatorKeys, key)
-		}
+		req.CoordinatorKeys = append(req.CoordinatorKeys, key)
 	}
 
 	splitReq := errors.Is(lastErr, errBrokerTooOld)

--- a/pkg/kgo/client.go
+++ b/pkg/kgo/client.go
@@ -1229,12 +1229,12 @@ func (cl *Client) loadCoordinators(ctx context.Context, typ int8, keys ...string
 					}
 				}
 			} else {
-				resp, _ := shard.Resp.(*kmsg.FindCoordinatorResponse)
+				resp := shard.Resp.(*kmsg.FindCoordinatorResponse)
 				for _, rc := range resp.Coordinators {
 					c, ok := key2load[rc.Key]
 					if ok {
 						c.err = kerr.ErrorForCode(rc.ErrorCode)
-						c.node = resp.NodeID
+						c.node = rc.NodeID
 					}
 				}
 			}

--- a/pkg/kgo/sink.go
+++ b/pkg/kgo/sink.go
@@ -329,7 +329,7 @@ func (s *sink) produce(sem <-chan struct{}) bool {
 			case isRetriableBrokerErr(err) || isDialErr(err):
 				s.cl.bumpRepeatedLoadErr(err)
 				s.cl.cfg.logger.Log(LogLevelWarn, "unable to AddPartitionsToTxn due to retriable broker err, bumping client's buffered record load errors by 1 and retrying", "err", err)
-				s.cl.triggerUpdateMetadata(false, "attempting to refresh transaction coordinator due to failed AddPartitionsToTxn requests")
+				s.cl.triggerUpdateMetadata(false, "attempting to refresh broker list due to failed AddPartitionsToTxn requests")
 				return moreToDrain || len(req.batches) > 0 // nothing stripped if request-issuing error
 			default:
 				// Note that err can be InvalidProducerEpoch, which is

